python/amici/ode_export.py

```diff
@@ -0,0 +1,2266 @@
+        ----------
+
+        """
+        if sign not in [-1, 1]:
+            raise TypeError(f'sign must be +1 or -1, was {sign}')
+
+        vars = dict()
+        for varname in [x, y]:
+            if var_in_function_signature(name, varname):
+                vars[varname] = self.sym(varname)
+            else:
+                vars[varname] = self.eq(varname)
+
+        if transpose_x:
+            xx = vars[x].transpose()
+        else:
+            xx = vars[x]
+
+        self._eqs[name] = sign * xx * vars[y]
+
+    def equationFromComponent(self, name, component):
+        """
+        Generates the formulas of a symbolic variable from the attributes
+
+        Arguments:
+        ----------
+        name: str
+            name of resulting symbolic variable
+
+        component: str
+            name of the attribute
+
+
+        Returns:
+        ----------
+
+        Raises:
+        ----------
+
+        """
+        self._eqs[name] = sp.DenseMatrix(
+            [comp.value for comp in getattr(self, component)]
+        )
+
+    def generateValue(self, name):
+        """
+        Generates the numeric values of a symbolic variable from value
+        prototypes
+
+        Arguments:
+        ----------
+        name: str
+            name of resulting symbolic variable
+
+        Returns:
+        ----------
+
+        Raises:
+        ----------
+
+        """
+        if name in self.value_prototype:
+            component = self.value_prototype[name]
+        else:
+            raise Exception(f'No values for {name}')
+
+        self._vals[name] = [comp.value for comp in getattr(self, component)]
+
+    def generateName(self, name):
+        """
+        Generates the names of a symbolic variable from variable prototypes or
+        equation prototypes
+
+        Arguments:
+        ----------
+        name: str
+            name of resulting symbolic variable
+
+        Returns:
+        ----------
+
+        Raises:
+        ----------
+
+        """
+        if name in self.variable_prototype:
+            component = self.variable_prototype[name]
+        elif name in self.equation_prototype:
+            component = self.equation_prototype[name]
+        else:
+            raise Exception(f'No names for {name}')
+
+        self._names[name] = [comp.name for comp in getattr(self, component)]
+
+def var_in_function_signature(name, varname):
+    """
+    Checks if the values for a symbolic variable is passed in the signature
+    of a function
+
+    Arguments:
+    ----------
+    name: str
+        name of the function
+
+    varname: str
+        name of the symbolic variable
+
+    Returns:
+    ----------
+
+    Raises:
+    ----------
+
+    """
+    varname = varname.replace('sparse', '')
+    return name in functions \
+           and re.search(
+                    f'const (realtype|double) \*{varname}[0]*[,)]+',
+                    functions[name]['signature']
+                )
+
+
+class ODEExporter:
```

Comment:
I propose separating C++ generation from compilation. Once there core files have been generated, there should be other functions/classes generating and compiling the interface files. This will me more modular and maintainable. 
It will allow people who just want to generate C++ code to not have swig installed. And if somebody feels like: It will also be easy to add another class for matlab wrapper generation, so that we can dump matlab sbml import completely.